app/utils/Query/AST.scala

```diff
@@ -1,23 +1,147 @@
 package utils.Query
 
+import play.api.Logger
 import play.api.libs.json.JsObject
+import slick.jdbc.{PositionedParameters, SQLActionBuilder, SetParameter}
+import slick.driver.MySQLDriver.api._
+import slick.profile.SqlAction
 
 import scala.util.parsing.input.Positional
 
 //Definition of the abstract syntax tree for querys
-sealed trait QueryAST extends Positional
-case class AndThen(step1: QueryAST, step2: QueryAST) extends QueryAST
+sealed trait QueryAST extends Positional {
+  //ToDo: function should return a SQLActionBuilder instance (SQL Injection)
+  //def toSqlStatement :String
+  def toSqlStatement : SQLActionBuilder
+}
+case class AndThen(step1: QueryAST, step2: QueryAST) extends QueryAST {
+  override def toSqlStatement = {
+    //step1.toSqlStatement + " AND " + step2.toSqlStatement
+    val sql1 = step1.toSqlStatement
+    val sql2 = step2.toSqlStatement
+
+    Converter.concat(sql1, Converter.concat(sql"""AND""", sql2))
+  }
+}
 
 sealed trait Combinations { def step: QueryAST }
-case class A(step1: QueryAST, step2: QueryAST) extends QueryAST //AND
-case class O(step1: QueryAST, step2: QueryAST) extends QueryAST //OR
+
+/**
+  * AND
+  * @param step1
+  * @param step2
+  */
+case class A(step1: QueryAST, step2: QueryAST) extends QueryAST {
+  override def toSqlStatement = {
+    val sql1 = step1.toSqlStatement
+    val sql2 = step2.toSqlStatement
+
+    Converter.concat(sql1, Converter.concat(sql""" AND """, sql2))
+  }
+}
+
+/**
+  * OR
+  * @param step1
+  * @param step2
+  */
+case class O(step1: QueryAST, step2: QueryAST) extends QueryAST {
+  override def toSqlStatement ={
+
+    val sql1 = step1.toSqlStatement
+    val sql2 = step2.toSqlStatement
+
+    Converter.concat(sql1, Converter.concat(sql""" OR """, sql2))
+  }
+}
+
+case class Combination(f1: QueryAST, f2: List[QueryParser.~[QueryToken, QueryAST]]) extends QueryAST {
+  override def toSqlStatement = {
+    var sql = f1.toSqlStatement
+    f2.foreach(o => {
+      sql = o._1 match{
+        case AND => Converter.concat(sql, Converter.concat(sql""" AND """, o._2.toSqlStatement))
+        case OR => Converter.concat(sql, Converter.concat(sql""" OR """, o._2.toSqlStatement))
+        case _ => throw new Exception
+      }
+    })
+    sql
+  }
+}
+
+case class Group(f1: QueryToken, f2: Combination, f3: QueryToken)extends QueryAST{
+  override def toSqlStatement = {
+    val sql = f2.toSqlStatement
+
+    Converter.concat(sql"""(""", Converter.concat(sql, sql""")"""))
+  }
+}
+
+case class CombinedGroup(f1: QueryAST, f2: List[QueryParser.~[QueryToken, QueryAST]]) extends QueryAST{
+  override def toSqlStatement = {
+    var sql = f1.toSqlStatement
+    f2.foreach(o => {
+      sql = o._1 match {
+        case AND => Converter.concat(sql, Converter.concat(sql""" AND """, o._2.toSqlStatement))
+        case OR => Converter.concat(sql, Converter.concat(sql""" OR """, o._2.toSqlStatement))
```

Comment:
@ottmann you believe yourself, that this can be removed :)

Reply:
I would not delete it yet, because it is the same code but handels different logical parts of the ast. 
Here is still more implementation work necessary to provide interleaved brackets